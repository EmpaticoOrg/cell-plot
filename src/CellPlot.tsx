import * as React from 'react';

import Frame from './Frame';
import Cells, {Coordinate} from './Cells';
import {PlotContext, PlotContextProps} from './PlotContext';
import {PlotConsumer, PlotConsumerProps} from './PlotConsumer';
import {convertCoordinate} from './util';

export interface Props {
  xLabels?: React.ReactNode[];
  verticalBorders?: string[]; // a cyclic sequence of strings for the css border property
  xs: number[];

  yLabels?: React.ReactNode[];
  horizontalBorders?: string[]; // a cyclic sequence of strings for the css border property
  ys: number[];

  onClick?: (x: number, y: number) => void; // TODO: convert to Coordinate in 2.x release
  onHover?: (coordinate: Coordinate) => void;
}

/**
 * Creates optionally clickable cells and a plottable coordinate plane.
 */
export default class CellPlot extends React.Component<Props> {
  static childContextTypes = {...PlotContextProps, ...PlotConsumerProps};
  el: HTMLElement;

  getChildContext(): PlotContext & PlotConsumer {
    const gridYMin = this.props.ys[0];
    const gridYMax = this.props.ys[this.props.ys.length - 1];
    const gridYStep = this.props.ys[1] ? this.props.ys[1] - gridYMin : 1;
    const gridYRange = gridYMax - gridYMin + gridYStep;

    const gridXMin = this.props.xs[0];
    const gridXMax = this.props.xs[this.props.xs.length - 1];
    const gridXStep = this.props.xs[1] ? this.props.xs[1] - gridXMin : 1;
    const gridXRange = gridXMax - gridXMin + gridXStep;

    return {
      gridXMin,
      gridXMax,
      gridXStep,
      gridXRange: gridXMax - gridXMin + gridXStep,
      gridYMin,
      gridYMax,
      gridYStep,
      gridYRange: gridYMax - gridYMin + gridYStep,
      yAt: (pxY: number): number => {
        const pxPerY = this.el.offsetHeight / gridYRange;
        return convertCoordinate(pxY / pxPerY, {min: gridYMin, max: gridYMax, interval: gridYStep});
      },
      xAt: (pxX: number): number => {
        const pxPerX = this.el.offsetWidth / gridXRange;
        return convertCoordinate(pxX / pxPerX, {min: gridXMin, max: gridXMax, interval: gridXStep});
      }
    };
  }

  render() {
    return (
      <Frame
        innerRef={(el) => this.el = el}
        xLabels={this.props.xLabels || []}
        yLabels={this.props.yLabels || []}
        verticalBorder={this.props.verticalBorders && this.props.verticalBorders[0]}
        horizontalBorder={this.props.horizontalBorders && this.props.horizontalBorders[0]}>

        <Cells
          xs={this.props.xs}
          verticalBorders={this.props.verticalBorders}
          ys={this.props.ys}
          horizontalBorders={this.props.horizontalBorders}
          onClick={this.props.onClick}
          onHover={this.props.onHover} />

        {this.props.children}
      </Frame>
    );
  }
}
