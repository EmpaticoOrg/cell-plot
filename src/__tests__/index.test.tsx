import * as React from 'react';
import * as renderer from 'react-test-renderer';

import CellPlot, {Rectangle} from '../index';

test('static render', () => {
  const component = renderer.create(
    <CellPlot
      xLabels={['x0', 'x1', 'x2']}
      yLabels={['y0', 'y1', 'y2']}
      xs={[-1, 0, 1]}
      ys={[-1, 0, 1]}>
      <Rectangle x={1} y={0} width={1} height={1} />
      <Rectangle x={0} y={1} width={1} height={1} />
      <Rectangle x={-1} y={0} width={1} height={1} />
      <Rectangle x={0} y={-1} width={1} height={1} />
    </CellPlot>
  );

  expect(component.toJSON()).toMatchSnapshot();
});

describe('layout calculations', () => {
  test('too far above', () => {
    const component = renderer.create(
      <CellPlot xs={[3, 4]} ys={[5, 6]}>
        <Rectangle x={3} y={4} width={1} height={1} />
      </CellPlot>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('too far right', () => {
    const component = renderer.create(
      <CellPlot xs={[3, 4]} ys={[5, 6]}>
        <Rectangle x={5} y={5} width={1} height={1} />
      </CellPlot>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('too far below', () => {
    const component = renderer.create(
      <CellPlot xs={[3, 4]} ys={[5, 6]}>
        <Rectangle x={3} y={7} width={1} height={1} />
      </CellPlot>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('too far left', () => {
    const component = renderer.create(
      <CellPlot xs={[3, 4]} ys={[5, 6]}>
        <Rectangle x={2} y={5} width={1} height={1} />
      </CellPlot>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('clamped by top boundary', () => {
    const component = renderer.create(
      <CellPlot xs={[3, 4]} ys={[5, 6]}>
        <Rectangle x={3} y={4} width={2} height={2} />
      </CellPlot>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('clamped by right boundary', () => {
    const component = renderer.create(
      <CellPlot xs={[3, 4]} ys={[5, 6]}>
        <Rectangle x={4} y={5} width={2} height={2} />
      </CellPlot>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('clamped by bottom boundary', () => {
    const component = renderer.create(
      <CellPlot xs={[3, 4]} ys={[5, 6]}>
        <Rectangle x={3} y={6} width={2} height={2} />
      </CellPlot>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('clamped by left boundary', () => {
    const component = renderer.create(
      <CellPlot xs={[3, 4]} ys={[5, 6]}>
        <Rectangle x={2} y={5} width={2} height={2} />
      </CellPlot>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

});
