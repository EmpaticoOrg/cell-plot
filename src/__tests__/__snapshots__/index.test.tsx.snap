// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`static render 1`] = `
<div
  className="sc-bdVaJa dqbzLu"
  style={
    Object {
      "flex": "1 1 auto",
    }
  }
>
  <div
    className="sc-bwzfXH kulSxP"
  >
    <div
      className="sc-ifAKCX eYukKq"
    >
      x0
    </div>
    <div
      className="sc-bxivhb hvlrXC"
    >
      y0
    </div>
    <div
      className="sc-bxivhb hvlrXC"
    >
      y1
    </div>
    <div
      className="sc-bxivhb hvlrXC"
    >
      y2
    </div>
  </div>
  <div
    className="sc-bxivhb hvlrXC"
  >
    <div
      className="sc-bwzfXH kulSxP"
      style={
        Object {
          "flex": "1 1 auto",
        }
      }
    >
      <div
        className="sc-bdVaJa dqbzLu"
      >
        <div
          className="sc-htpNat kBMfje"
        >
          x0
        </div>
        <div
          className="sc-htpNat kBMfje"
        >
          x1
        </div>
        <div
          className="sc-htpNat kBMfje"
        >
          x2
        </div>
      </div>
      <div
        className="sc-bxivhb hvlrXC"
        style={
          Object {
            "borderLeft": undefined,
            "borderTop": undefined,
          }
        }
      >
        <div
          style={
            Object {
              "flex": "1 1 auto",
              "position": "relative",
            }
          }
        >
          <div
            className="sc-EHOje eoRSaj"
          >
            <div
              className="sc-bZQynM loNdad"
              style={
                Object {
                  "borderRight": undefined,
                }
              }
            >
              <div
                className="sc-gzVnrw gfFhBZ"
                onClick={[Function]}
                style={
                  Object {
                    "borderBottom": undefined,
                  }
                }
              />
              <div
                className="sc-gzVnrw gfFhBZ"
                onClick={[Function]}
                style={
                  Object {
                    "borderBottom": undefined,
                  }
                }
              />
              <div
                className="sc-gzVnrw gfFhBZ"
                onClick={[Function]}
                style={
                  Object {
                    "borderBottom": undefined,
                  }
                }
              />
            </div>
            <div
              className="sc-bZQynM loNdad"
              style={
                Object {
                  "borderRight": undefined,
                }
              }
            >
              <div
                className="sc-gzVnrw gfFhBZ"
                onClick={[Function]}
                style={
                  Object {
                    "borderBottom": undefined,
                  }
                }
              />
              <div
                className="sc-gzVnrw gfFhBZ"
                onClick={[Function]}
                style={
                  Object {
                    "borderBottom": undefined,
                  }
                }
              />
              <div
                className="sc-gzVnrw gfFhBZ"
                onClick={[Function]}
                style={
                  Object {
                    "borderBottom": undefined,
                  }
                }
              />
            </div>
            <div
              className="sc-bZQynM loNdad"
              style={
                Object {
                  "borderRight": undefined,
                }
              }
            >
              <div
                className="sc-gzVnrw gfFhBZ"
                onClick={[Function]}
                style={
                  Object {
                    "borderBottom": undefined,
                  }
                }
              />
              <div
                className="sc-gzVnrw gfFhBZ"
                onClick={[Function]}
                style={
                  Object {
                    "borderBottom": undefined,
                  }
                }
              />
              <div
                className="sc-gzVnrw gfFhBZ"
                onClick={[Function]}
                style={
                  Object {
                    "borderBottom": undefined,
                  }
                }
              />
            </div>
          </div>
          <div
            className="sc-htoDjs btzGSf"
            style={
              Object {
                "height": "33.33333333333333%",
                "left": "66.66666666666666%",
                "top": "33.33333333333333%",
                "width": "33.33333333333333%",
              }
            }
          />
          <div
            className="sc-htoDjs btzGSf"
            style={
              Object {
                "height": "33.33333333333333%",
                "left": "33.33333333333333%",
                "top": "66.66666666666666%",
                "width": "33.33333333333333%",
              }
            }
          />
          <div
            className="sc-htoDjs btzGSf"
            style={
              Object {
                "height": "33.33333333333333%",
                "left": "0%",
                "top": "33.33333333333333%",
                "width": "33.33333333333333%",
              }
            }
          />
          <div
            className="sc-htoDjs btzGSf"
            style={
              Object {
                "height": "33.33333333333333%",
                "left": "33.33333333333333%",
                "top": "0%",
                "width": "33.33333333333333%",
              }
            }
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
